src/analytics-preload.ts

```diff
@@ -0,0 +1,248 @@
+  }
+
+  public get list(): PreInitMethodCall[] {
+    return [...this._list, ...this._windowBuffer]
+  }
+
+  push(...calls: PreInitMethodCall[]) {
+    this._list.push(...calls)
+  }
+
+  clear(): void {
+    this._list = []
+  }
+
+  /**
+   *  Call method and mark as "called"
+   */
+  async callMethod<T extends PreInitMethodName>(
+    analytics: Analytics,
+    methodCall: PreInitMethodCall<T>
+  ) {
+    if (methodCall.called) {
+      return undefined
+    }
+
+    methodCall.called = true
+
+    const { method, args, resolve } = methodCall
+    const result = await (analytics[method] as Function)(...args)
+    resolve(result)
+  }
+}
+
+type AnalyticsLoader = (
+  preInitBuffer: PreInitMethodCallBuffer
+) => Promise<[Analytics, Context]>
+
+export class AnalyticsBuffered implements PromiseLike<[Analytics, Context]> {
+  instance?: Analytics
+  ctx?: Context
+
+  private preInitBuffer = new PreInitMethodCallBuffer()
+  private onFulfilled?: (instance: [Analytics, Context]) => unknown
+  private onRejected?: (err: unknown) => unknown
+
+  constructor(loader: AnalyticsLoader) {
+    const promise = loader(this.preInitBuffer)
+    void this.run(promise)
+  }
+
+  // register onFulfilled and onRejected methods on class.
+  // while AnalyticsWithBuffer instance is only promise-like,
+  // .then should return a Promise<[Analytics...]> and not another AnalyticsWithBuffer
+  then<T, T2 = never>(
+    onFulfilled: (instance: [Analytics, Context]) => T | PromiseLike<T>,
+    onRejected?: (reason: unknown) => T2 | PromiseLike<T2>
+  ) {
+    return new Promise<T>((resolve, reject) => {
+      this.onFulfilled = (instance: [Analytics, Context]) => {
+        try {
+          const result = onFulfilled(instance)
+          resolve(result)
+        } catch (err) {
+          reject(err)
+        }
+      }
+
+      if (onRejected) {
+        this.onRejected = (err: unknown) => {
+          try {
+            if (onRejected) {
+              reject(onRejected(err))
+            }
+          } catch (err) {
+            reject(err)
+          }
+        }
+      }
+    })
+  }
+
+  catch<T>(onRejected: (reason: unknown) => T | PromiseLike<T>) {
+    return new Promise<T>((_, reject) => {
+      this.onRejected = (err: unknown) => {
+        try {
+          if (onRejected) {
+            reject(onRejected(err))
+          }
+        } catch (err) {
+          reject(err)
+        }
+      }
+    })
+  }
+
+  async run(analyticsPromise: Promise<[Analytics, Context]>) {
+    try {
+      const [analytics, context] = await analyticsPromise
+      if (this.onFulfilled) {
+        this.onFulfilled([analytics, context])
+      }
+      this.instance = analytics
+      this.ctx = context
+      return [analytics, context]
+    } catch (err) {
+      if (this.onRejected) {
+        this.onRejected(err)
+      } else {
+        throw err
+      }
+    }
+  }
+
+  private _createMethod<T extends PreInitMethodName>(methodName: T) {
```

Comment:
Does this make every method an async method? I'm trying to think about if our synchronous methods are incapable of throwing errors today. If they start returning promises then the way users handle errors will have to change as well since a try/catch will only work if they await the function. That could be a breaking change.

Reply:
Every method is technically an async method because we really have no idea when the method is going to be called -- which is why we store the "reference" to resolve in the class.

This is an upgrade over the snippet API, where the type signature was basically a lie -- whether or not window.analytics.track returns a promise has to do with where it's called in our lifecycle -- which seems way more error prone than this.

Reply:
Totally agree in the value of the APIs that currently return a promise returning a promise even when they're buffered.

It looks like of the methods that don't return a promise are:
- reset
  - returns void already
- debug
  - returns `Analytics` so could return `AnalyticsBuffered`
- on/off/once
  - returns `Analytics` so could return `AnalyticsBuffered`. This one feels the most weird to make promise-based since event emitters are so common in JS and it's expected you can just chain `.on()` calls.  
- setAnonymousId
  - returns `ID`. This is the trickiest one. _Technically_ ID can return undefined - we could document that it returns undefined if the instance hasn't loaded yet which is today's behavior on the CDN. I'm not sure what to do for this method - if we returned void from the buffered call like we do today with the snippet the API is still unchanged - maybe it is just a matter of documenting the behavior.